### PR TITLE
raftstore: check merge entries' length to avoid slice out of range

### PR DESCRIPTION
**Pull request by Connor1996**
## What have you changed? (mandatory)
In merge process, target region will send entries to fall-behind source region. It is possible to propose a `CompactLog` command after proposing `PrepareMerge`.  Note that, the `entries` involved in `CommitMerge` would not take the `CompactLog` record.

So during the merge process, if the source region catches up with the logs by raft append msgs, `self.raft_group.raft.raft_log.committed - log_idx` would be larger than `entries.len()` which leads to `entries[self.raft_group.raft.raft_log.committed - log_idx..]` out of slice range. 

So check the len before cutting up the slice.

## What are the type of changes? (mandatory)
- Bugfix (a change which fixes an issue)

## How has this PR been tested? (mandatory)
unit-test

## Does this PR affect documentation (docs) or release note? (mandatory)
No

## Does this PR affect `tidb-ansible` update? (mandatory)
No

## Refer to a related PR or issue link (optional)
https://github.com/pingcap/tidb/issues/11744

### CODE REVIEW

**Comment by Connor1996** [comment]
PTAL @BusyJay @overvenus @hicqu

File: tests/failpoints/cases/test_merge.rs [hunk 1]
```diff
@@ -308,6 +308,79 @@ fn test_node_merge_catch_up_logs_leader_election() {
...
+    let peer_on_store1 = find_peer(&region, 1).unwrap().to_owned();
+    cluster.must_transfer_leader(region.get_id(), peer_on_store1);
+    cluster.must_split(&region, b"k2");
+    let left = pd_client.get_region(b"k1").unwrap();
+    let right = pd_client.get_region(b"k2").unwrap();
+
+    // put some keys to trigger compact raft log
+    for i in 2..20 {
+        cluster.must_put(format!("k1{}", i).as_bytes(), b"v");
+        must_get_none(&cluster.get_engine(3), b"k1{}");
```
Comment (overvenus) [inline tests/failpoints/cases/test_merge.rs]: Wrong key?

File: tests/failpoints/cases/test_merge.rs [hunk 2]
```diff
@@ -308,6 +308,79 @@ fn test_node_merge_catch_up_logs_leader_election() {
...
+    thread::sleep(Duration::from_millis(100));
+
+    // let source region not merged
+    fail::cfg("on_handle_catch_up_logs_for_merge", "pause").unwrap();
+    fail::cfg("after_handle_catch_up_logs_for_merge", "pause").unwrap();
+    // due to `on_handle_catch_up_logs_for_merge` failpoint, we already pass `apply_index < catch_up_logs.merge.get_commit()`
+    // so now can let apply index make progress.
+    fail::remove("apply_after_prepare_merge");
+
+    // make sure all the logs are committed, including the compact command
```
Comment (overvenus) [inline tests/failpoints/cases/test_merge.rs]: How to make sure logs are compacted? What about checking the truncated index?

File: src/raftstore/store/fsm/peer.rs [hunk 3]
```diff
@@ -1862,19 +1862,20 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
         self.fsm.peer.pending_merge_state = Some(state);
         self.notify_prepare_merge();
 
         if let Some(logs_up_to_date) = self.fsm.peer.catch_up_logs.take() {
-            logs_up_to_date.store(region.get_id(), Ordering::SeqCst);
-            let mailbox = self.ctx.apply_router.mailbox(target).unwrap();
             // Send CatchUpLogs back to destroy source apply delegate,
             // then it will send `LogsUpToDate` to target apply delegate.
+            let mut req = CommitMergeRequest::new();
+            req.set_commit(commit);
```
Comment (overvenus) [inline src/raftstore/store/fsm/peer.rs]: Is it necessary to attach commit?

File: src/raftstore/store/fsm/apply.rs [hunk 4]
```diff
@@ -1805,26 +1799,17 @@ impl ApplyDelegate {
             );
 
             // Sends message to the source apply worker and pause `exec_commit_merge` process
-            if let Some(mailbox) = ctx.router.mailbox(self.region_id()) {
```
Comment (BusyJay) [inline src/raftstore/store/fsm/apply.rs]: Is it possible that when the task is sent back, the mailbox is changed? For example, target peer is removed and readded.

File: src/raftstore/store/fsm/apply.rs [hunk 5]
```diff
@@ -2547,27 +2531,25 @@ impl ApplyFsm {
         }
 
         // if it is already up to date, no need to catch up anymore
-        if catch_up_logs.logs_up_to_date.load(Ordering::SeqCst) == 0 {
```
Comment (BusyJay) [inline src/raftstore/store/fsm/apply.rs]: Why skip the check now?

File: src/raftstore/store/fsm/apply.rs [hunk 6]
```diff
@@ -2547,27 +2531,25 @@ impl ApplyFsm {
...
+                region_id: self.delegate.region_id(),
+                apply_state: self.delegate.apply_state.clone(),
+                exec_res: res,
+                metrics: self.delegate.metrics.clone(),
+                applied_index_term: self.delegate.applied_index_term,
+            });
+            return;
         }
+
+        fail_point!("after_handle_catch_up_logs_for_merge");
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/raftstore/store/fsm/apply.rs]: No, need to block other peers too

File: tests/failpoints/cases/test_merge.rs (hunk 2 above)
Comment (Connor1996) [reply tests/failpoints/cases/test_merge.rs]: The compact log won't actually be executed, so checking the truncated index is useless.

File: src/raftstore/store/fsm/peer.rs (hunk 3 above)
Comment (Connor1996) [reply src/raftstore/store/fsm/peer.rs]: Not necessary, but it's better to take it to double-check `apply_index >= commit`

File: src/raftstore/store/fsm/apply.rs (hunk 5 above)
Comment (Connor1996) [reply src/raftstore/store/fsm/apply.rs]: note that we don't set `logs_up_to_date` in `on_ready_prepare_merge` anymore, it is updated only after `apply_index >= catch_up_logs.merge.get_commit()`. so here `logs_up_to_date` is always zero

File: tests/failpoints/cases/test_merge.rs (hunk 1 above)
Comment (Connor1996) [reply tests/failpoints/cases/test_merge.rs]: we don't care the value of these keys, so just set `v`

File: src/raftstore/store/fsm/apply.rs (hunk 4 above)
Comment (Connor1996) [reply src/raftstore/store/fsm/apply.rs]: Even that, seems it's okay to send `LogsUpToDate` to newly readded target peer, cause `LogsUpToDate` is just an empty message which is only used to trigger target peer to check `resume_pending_merge`, so it won't make any effect on the target peer.

File: tests/failpoints/cases/test_merge.rs (hunk 1 above)
Comment (overvenus) [reply tests/failpoints/cases/test_merge.rs]: I think this key `b"k1{}"` is wrong, or please remove the `{}`.

Comment (Connor1996) [reply tests/failpoints/cases/test_merge.rs]: Oh, by mistake.

**Comment by BusyJay** [comment]
I remove the key fix in src/raftstore/store/peer.rs, the test can still pass.

**Comment by Connor1996** [comment]
@BusyJay Here is what the result shows when I comment the key fix, and it can be reproduced everytime on idc. Maybe should enlarge the wait time to trigger compact?
```
running 1 test
thread 'raftstore-3-0::test_node_merge_catch_up_logs_no_need' panicked at 'slice index starts at 3 but ends at 2', src/libcore/slice/mod.rs:2561:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
thread 'apply-0::test_node_merge_catch_up_logs_no_need' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', src/libcore/result.rs:1051:5
thread 'cases::test_merge::test_node_merge_catch_up_logs_no_need' panicked at 'called `Result::unwrap()` on an `Err` value: Any', src/libcore/result.rs:1051:5
test cases::test_merge::test_node_merge_catch_up_logs_no_need ... FAILED
```

**Comment by Connor1996** [comment]
@BusyJay I try it on my poor laptop, and it also can be reproduced steadily. How you test it?

**Pull request merged**